### ISSUE TITLE
Tikv crash in compaction filter casued by IoError: No Such file or directory: While open a file for random read: /data/xxx/yyy/zzz/115748.blob

### ISSUE THREAD

**Issue by mzygQAQ** [tikv/titan#328]
This error is different with the existing older Missing blob.

**Comment by mzygQAQ** [comment]
```
Status BlobStorage::Get(const ReadOptions& options, const BlobIndex& index,
                        BlobRecord* record, PinnableSlice* buffer) {
  auto sfile = FindFile(index.file_number).lock();
  if (!sfile)
    return Status::Corruption("Missing blob file: " +
                              std::to_string(index.file_number));
 
  // sfile is not nullptr, but error in the following line
  return file_cache_->Get(options, sfile->file_number(), sfile->file_size(),
                          index.blob_handle, record, buffer);
}
```

**Comment by mzygQAQ** [comment]
I have checked the code here and there is indeed a race condition present

**Comment by mzygQAQ** [comment]
@v01dstar Hello, can you help confirm

**Comment by v01dstar** [comment]
Are you able to reproduce this error in TiKV?

I believe that `purge obesolete` will only delete files that are guaranteed not to be referenced by any keys in SSTs. This is done by giving a file to be purged a sequence number after the references are updated in SSTs, and only when the assigned sequence number is the `OldestSequenceNumber`, then obesolete files can be purged. i.e.

Blob File GC starts ->  Genereate new files -> Rewrite references in SSTs -> Get the sequence number -> Delete input files (mark tombstone, not physically delete, but mark deleting with the sequence number) -> Another job will periodically scan obsolete files who have sequence number older than the current oldest snapshot .

So, in theory. Any read with a snapshot should not acess data in deleted files.

**Comment by v01dstar** [comment]
At first glance, seems possible, allow me dig more.

**Comment by v01dstar** [comment]
I think this is indeed a problem, unless we set `skip_value_in_compaction_filter` to be true, however, we don't. I am surprise that we don't see this error in our users' environment. If I didn't miss anything, this is more than a race condition. Since compaction filter does not go through the normal read path (i.e. read with a snapshot), this should happen quite frequently.
@Connor1996 , do you know why this was not found in Titan's use cases?

**Comment by mzygQAQ** [comment]
I guess that in the TIDB environment, TikTok only uses Compaction Filter in WriteCF, while WriteCF only saves some transaction commit information and small values less than 256 bytes. Moreover, by default, WriteCF does not enable Titan, so it will not occur.
This issue occurs in scenarios where Tikv is used with Rawkv or directly with Titan.

**Comment by v01dstar** [comment]
> I guess that in the TIDB environment, Tikv only uses Compaction Filter in WriteCF, while WriteCF only saves some transaction commit information and small values less than 256 bytes. Moreover, by default, WriteCF does not enable Titan, so it will not occur. This issue occurs in scenarios where Tikv is used with Rawkv or directly with Titan.

Yes, I totally missed that. I guess, you can leverate `skip_value_in_compaction_filter` in this case. Or you can propose a simple fix, which as you suggested, and also mentioned in the TODO, i.e. return corresponding error to the caller of `Get()`, and the caller (compaction filter) decide what to do.

**Comment by mzygQAQ** [comment]
i try
